cmd/adapter/config.go

```diff
@@ -0,0 +1,272 @@
+		SyncIntervalSec int
+	}
+)
+
+func LoadCfgFromFile(path string) (*Config, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	cfg := &Config{}
+	err = json.Unmarshal(data, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+//===================== config =====================
+
+func NewDefaultConfig() *Config {
+	return &Config{
+		Gravity: &gravity{
+			ApiListenAddr: "127.0.0.1:8083",
+			Kubernetes: &kube{
+				Namespace:              "kube-system",
+				ForwarderConfigMapName: "log-forwarders",
+			},
+		},
+
+		Logrange: &logrange{
+			Partition:         "logrange.pipe=__default__",
+			ForwarderTmplFile: "/opt/logrange/gravity/config/forward-tmpl.json",
+			Kubernetes: &kube{
+				Namespace:              "kube-system",
+				ForwarderConfigMapName: "lr-forwarder",
+			},
+			Transport: &transport.Config{
+				ListenAddr: "127.0.0.1:9966",
+			},
+		},
+
+		SyncIntervalSec: 5,
+	}
+}
+
+func (c *Config) Apply(other *Config) {
+	if other == nil {
+		return
+	}
+
+	if other.Gravity != nil {
+		c.Gravity.apply(other.Gravity)
+	}
+	if other.Logrange != nil {
+		c.Logrange.apply(other.Logrange)
+	}
+	if other.SyncIntervalSec != 0 {
+		c.SyncIntervalSec = other.SyncIntervalSec
+	}
+}
+
+func (c *Config) check() error {
+	if c.Gravity == nil {
+		return fmt.Errorf("invalid Gravity=%v, must be non-nil", c.Gravity)
+	}
+	if c.Logrange == nil {
+		return fmt.Errorf("invalid Logrange=%v, must be non-nil", c.Logrange)
+	}
+	if err := c.Gravity.check(); err != nil {
+		return fmt.Errorf("invalid Gravity=%v: %v", c.Gravity, err)
+	}
+	if err := c.Logrange.check(); err != nil {
+		return fmt.Errorf("invalid Logrange=%v: %v", c.Logrange, err)
+	}
+	if c.SyncIntervalSec <= 0 {
+		return fmt.Errorf("invalid SyncIntervalSec=%v, must be > 0sec", c.SyncIntervalSec)
+	}
+	return nil
+}
+
+func (c *Config) String() string {
+	return utils.ToJsonStr(c)
+}
+
+//===================== gravity =====================
+
+func (g *gravity) apply(other *gravity) {
+	if other == nil {
+		return
+	}
+
+	if other.Kubernetes != nil {
+		g.Kubernetes.apply(other.Kubernetes)
+	}
+	if other.ApiListenAddr != "" {
+		g.ApiListenAddr = other.ApiListenAddr
+	}
+}
+
+func (g *gravity) check() error {
+	if g.Kubernetes.check() == nil {
+		return fmt.Errorf("invalid Kubernetes=%v: must be non-nil", g.Kubernetes)
+	}
+	if err := g.Kubernetes.check(); err != nil {
+		return fmt.Errorf("invalid Kubernetes=%v: %v", g.Kubernetes, err)
+	}
+	if g.ApiListenAddr == "" {
+		return fmt.Errorf("invalid ApiListenAddr=%v, must be non-empty", g.ApiListenAddr)
+	}
+	return nil
+}
+
+func (g *gravity) String() string {
+	return utils.ToJsonStr(g)
+}
+
+//===================== logrange =====================
+
+func (l *logrange) apply(other *logrange) {
+	if other == nil {
+		return
+	}
+
+	if other.Transport != nil {
+		l.Transport.Apply(other.Transport)
+	}
+	if other.Kubernetes != nil {
+		l.Kubernetes.apply(other.Kubernetes)
```

Comment:
Looking at these config structs and their methods, I feel like with so many config fields being pointers and the lack of nil checks, something somewhere will definitely panic :) Could you please make sure all nils are handled properly and either add explicit nil checks (so you don't have to look into the actual method to see if it handles nil properly) or add tests for these cases?

Reply:
as I mentioned check() method on configs does this check, also I added some tests, applied (not yet merged to this branch)